### PR TITLE
Refactor host v2

### PR DESCRIPTION
**Pull request by gavin-ygy**
Refactor the host for each guest.

### CODE REVIEW

File: sdk/src/local/prover.rs [hunk 1]
```diff
@@ -66,18 +68,20 @@ impl ProverTask {
 
 pub struct LocalProver {
     tasks: Arc<Mutex<HashMap<String, Arc<Mutex<ProverTask>>>>>,
+    vk_path: String,
 }
 
 impl Default for LocalProver {
     fn default() -> Self {
-        Self::new()
+        Self::new("")
```
Comment (eigmax) [inline sdk/src/local/prover.rs:76]: If it's empty by default, the vk_path should be of optional

Comment (eigmax) [reply sdk/src/local/prover.rs:76]: Or we must remove the Default trait for Local Prover.

File: sdk/src/local/prover.rs [hunk 2]
```diff
@@ -119,6 +124,30 @@ impl Prover for LocalProver {
...
+        //let inputdir = format!("{}/input", vk_path);
+        fs::create_dir_all(vk_path).unwrap();
+        let should_agg = crate::local::stark::prove_stark(input, vk_path, &mut result).unwrap();
+        if !should_agg {
+            log::info!("Setup: generating the stark proof false, please check the SEG_SIZE or other parameters.");
+            return Err(anyhow::anyhow!(
+                "Setup: generating the stark proof false, please check the SEG_SIZE or other parameters!"));
+        }
+
+        if crate::local::snark::setup(vk_path) {
```
Comment (eigmax) [inline sdk/src/local/prover.rs:143]: match xxxx::setup(xxx) {
     true => Ok(()),
     _ => bail!("snark setup false")
}

File: sdk/src/local/snark.rs [hunk 3]
```diff
@@ -19,3 +20,16 @@ pub fn prove_snark(inputdir: &str, outputdir: &str) -> bool {
 pub fn prove_snark(inputdir: &str, outputdir: &str) -> bool {
     return false;
 }
+
+#[cfg(feature = "snark")]
+pub fn setup(inputdir: &str) -> bool {
+    let inputdir = std::ffi::CString::new(inputdir).unwrap();
```
Comment (eigmax) [inline sdk/src/local/snark.rs:26]: why does the setup return a bool?  
Rust is not C, use Result as possible as you can.  

change the prove_snark too.

File: sdk/src/local/snark.rs [hunk 4]
```diff
@@ -19,3 +20,16 @@ pub fn prove_snark(inputdir: &str, outputdir: &str) -> bool {
...
+
+    let ret = unsafe { Setup(inputdir.as_ptr()) };
+    ret == 0
+}
+
+#[cfg(not(feature = "snark"))]
+pub fn setup(inputdir: &str) -> bool {
+    return false;
```
Comment (eigmax) [inline sdk/src/local/snark.rs:34]: why it return false?  just propagate a unimplemented panic!

Comment (gavin-ygy) [reply sdk/src/local/snark.rs:34]: the libsnark exposes  C  funtion for users

File: sdk/src/local/snark.rs (hunk 3 above)
Comment (gavin-ygy) [reply sdk/src/local/snark.rs:26]: the libsnark exposes C funtion for users

Comment (eigmax) [reply sdk/src/local/snark.rs:26]: 你这个程序有几种返回方式？？？？  你知道为什么Rust要引入Result吗？

Comment (gavin-ygy) [reply sdk/src/local/snark.rs:26]: Done

File: sdk/src/local/snark.rs (hunk 4 above)
Comment (gavin-ygy) [reply sdk/src/local/snark.rs]: Done

File: sdk/src/lib.rs [hunk 5]
```diff
@@ -4,42 +4,288 @@ pub mod prover;
...
+    pub async fn network(client_type: &ClientType) -> Self {
         Self {
-            prover: Box::new(NetworkProver::new().await.unwrap()),
+            prover: Box::new(NetworkProver::new(client_type).await.unwrap()),
+        }
+    }
+
+    //If the vk or pk doesn't exist, it will run setup().
+    pub async fn setup(&self, zkm_prover: &str, vk_path: &str, prover_input: &ProverInput) {
+        if zkm_prover.to_lowercase() == *LOCAL_PROVER {
```
[the remark this answers is not in the document]
Comment (gavin-ygy) [reply sdk/src/lib.rs:69]: Done

[the remark this answers is not in the document]
Comment (gavin-ygy) [reply sdk/src/lib.rs]: It has  pub struct ProverClient {
    pub prover: Box<dyn Prover>,
} in the file.   The ClientType  mainly  is new client (local or network).

File: sdk/src/network/prover.rs [hunk 6]
```diff
@@ -1,12 +1,12 @@
...
 use stage_service::stage_service_client::StageServiceClient;
 use stage_service::{GenerateProofRequest, GetStatusRequest};
-use std::env;
+
 use std::time::Instant;
 use tonic::transport::Endpoint;
 use tonic::transport::{Channel, ClientTlsConfig};
 
-use crate::prover::{Prover, ProverInput, ProverResult};
+use crate::prover::{ClientType, Prover, ProverInput, ProverResult};
```
[the remark this answers is not in the document]
Comment (gavin-ygy) [reply sdk/src/network/prover.rs:9]: ProverClient has been used in the lib.rs .
pub struct ProverClient {
    pub prover: Box<dyn Prover>,
}

File: host-program/mem-alloc-vec/src/main.rs [hunk 7]
```diff
@@ -0,0 +1,125 @@
...
+    let prover_client = ProverClient::new(&client_config).await;
+    log::info!("new prover client,ok.");
+
+    let prover_input = ProverInput {
+        elf: read(elf_path).unwrap(),
+        public_inputstream: vec![],
+        private_inputstream: vec![],
+        seg_size: seg_size1,
+        execute_only: execute_only1,
+        args: "".into(),
```
Comment (eigmax) [inline host-program/mem-alloc-vec/src/main.rs:74]: 这个args跟set guest input有啥关系? 直接去掉这个args!!!

File: host-program/revme/src/main.rs [hunk 8]
```diff
@@ -0,0 +1,128 @@
...
+use std::time::Instant;
+use zkm_sdk::{prover::ClientCfg, prover::ProverInput, ProverClient};
+
+pub const DEFAULT_PROVER_NETWORK_RPC: &str = "https://152.32.186.45:20002";
+pub const DEFALUT_PROVER_NETWORK_DOMAIN: &str = "stage";
+
+#[tokio::main]
+async fn main() -> Result<(), Box<dyn std::error::Error>> {
+    env_logger::try_init().unwrap_or_default();
+    /*let args: Vec<String> = env::args().collect();
```
Comment (eigmax) [inline host-program/revme/src/main.rs:13]: 没用的代码去掉!!!!!

**Review by eigmax — APPROVED** [review]

**Pull request merged**